### ISSUE TITLE
[Question][Bug]: Similarity Score and Order in PGVectorStore

### ISSUE THREAD

**Issue by rendyfebry** [jerryjliu/llama_index#7214]
### Question Validation

- [X] I have searched both the documentation and discord for an answer.

### Question

1. As per my understanding, the `score` on `NodeWithScore` indicates the similarity between the node content and the query. The score itself is in the range from 0 to 1, the higher the number, the more similar they are. (This is only my assumption based on this snippet, please CMIIW)
```
nodes = [
  NodeWithScore(node=Node(text="text"), score=0.7),
  NodeWithScore(node=Node(text="text"), score=0.8)
]

# filter nodes below 0.75 similarity score
processor = SimilarityPostprocessor(similarity_cutoff=0.75)
filtered_nodes = processor.postprocess_nodes(nodes)
```

2. If the above assumption is correct, then when querying the index, we expect to get nodes with scores ordered from **HIGHER TO LOWER**.

3. On `PGVectorStore`, we [query](https://github.com/jerryjliu/llama_index/blob/main/llama_index/vector_stores/postgres.py#L170...L172) with a simple `.order_by()` clause, which results in ascending order, so the scores would be ordered from **LOWER TO HIGHER**.
```
  stmt = select(  # type: ignore
      self.table_class,
      self.table_class.embedding.l2_distance(embedding)
  ).order_by(self.table_class.embedding.l2_distance(embedding))
```
4. But if we look further, the query uses the [l2_distance function from PgVector](https://github.com/pgvector/pgvector#vector-functions) which returns the Euclidean Distance. I'm not familiar with this algorithm, but from the name "distance" I assume shorter distance means higher similarity. So from the Euclidean Distance pov, we actually expect the result to be sorted by l2_distances from **LOWER TO HIGHER**.

![Euclidean Distance](https://github.com/jerryjliu/llama_index/assets/1105460/2763025d-2f1a-4a48-b6ec-d827f344dca8)

---

TL: DR So at the end the order of the results is still as expected, it returns nodes from the most similar to least similar (shorter l2_distance to the higher ones), but since the `PGVectorStore` uses Euclidean Distance as the similarity score, the similarities score order is reversed, from smaller to higher, which will mess up the NodePostProcessor logic.

Is my understanding correct? If yes, can we fix the similarity score calculation on the `PGVectorStore`?

**Comment by rendyfebry** [comment]
@dosu-beta

The problem is the PGVectorStore uses the Euclidean distance number as the similarity score, as you can see in the `PGVectorStore._query_with_score` function

```python
    def _query_with_score(
        self,
        embedding: Optional[List[float]],
        limit: int = 10,
        metadata_filters: Optional[MetadataFilters] = None,
    ) -> List[DBEmbeddingRow]:
        stmt = self._build_query(embedding, limit, metadata_filters)
        with self._session() as session:
            with session.begin():
                res = session.execute(stmt)
                return [
                    DBEmbeddingRow(
                        node_id=item.node_id,
                        text=item.text,
                        metadata=item.metadata_,
                        similarity=sim,
                    )
                    for item, sim in res.all()
                ]
```

Due to this logic, the first returned node will have the lowest score because it uses the Euclidean distance as the similarity score. This is not expected, because when we query the index, we expect to get the node with the highest score first.

**Comment by logan-markewich** [comment]
@rendyfebry hmmm, just tried out a few queries.

I don't think it's as simple as it seems. Most queries seem to return a "similarity" of around 0.5-0.6. Like copy/pasting an exact sentence and querying with that returns 0.5

Are the any details you could find about how l2 is actually calculated? We just need to normalize it to a scale of 0-1

**Comment by logan-markewich** [comment]
@rendyfebry this seems to work better? But the similarity is now inverted 

stmt = select(  # type: ignore
            self.table_class, self.table_class.embedding.cosine_distance(embedding)
        ).order_by(self.table_class.embedding.cosine_distance(embedding))

**Comment by rendyfebry** [comment]
Thanks' @logan-markewich 

I'm not sure whether to use the `Euclidean Distance` or `Cosine Distance`, from a quick Google search, both seem to work.

But the problem is actually on the scoring scale, since both are the distance the scale is still from 0 to 1, where a smaller number indicates better similarity. So we still need to fix the score.
```
similarity = ( 1 - distance )
```

I see your PR already covers that, which is good. And turns out the same logic actually used on other VectorStore also:
- https://github.com/jerryjliu/llama_index/blob/main/llama_index/vector_stores/chroma.py#L167
- https://github.com/jerryjliu/llama_index/blob/main/llama_index/vector_stores/supabase.py#L166

Anyway, I made a small PR to improve the clarity:
https://github.com/jerryjliu/llama_index/pull/7226
